File: jimkv/server/test/helper/table.cpp
// Copyright 2019 The JimDB Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied. See the License for the specific language governing
// permissions and limitations under the License.

#include "table.h"

namespace jimkv {
namespace test {
namespace helper {

Table::Table(const std::string& table_name, uint32_t table_id) {
    meta_.set_db_id(default_db_id);
//    meta_.set_db_name("shark_test_db");
    meta_.set_name(table_name);
    meta_.set_id(table_id);
}

Table::Table(const basepb::TableInfo& meta) : meta_(meta) {
    if (!meta_.columns().empty()) {
        last_col_id_ = meta_.columns(meta_.columns_size() - 1).id();
    }
}

std::vector<basepb::ColumnInfo> Table::GetPKs() const {
    std::vector<basepb::ColumnInfo> cols;
    for (const auto& col : meta_.columns()) {
        if (col.primary() > 0) cols.push_back(col);
    }
    return cols;
}

std::vector<basepb::ColumnInfo> Table::GetNonPkColumns() const {
    std::vector<basepb::ColumnInfo> cols;
    for (const auto& col : meta_.columns()) {
        if (col.primary() == 0) cols.push_back(col);
    }
    return cols;
}

std::vector<basepb::ColumnInfo> Table::GetAllColumns() const {
    std::vector<basepb::ColumnInfo> cols;
    for (const auto& col : meta_.columns()) {
        cols.push_back(col);
    }
    return cols;
}

basepb::ColumnInfo Table::GetColumn(uint64_t id) const {
    auto it = std::find_if(meta_.columns().cbegin(), meta_.columns().cend(),
            [id](const basepb::ColumnInfo& col){ return col.id() == id; });
    if (it != meta_.columns().cend()) {
        return *it;
    } else {
        throw std::runtime_error(std::string("column id not found: ") + std::to_string(id));
    }
}

basepb::ColumnInfo Table::GetColumn(const std::string& name) const {
    auto it = std::find_if(meta_.columns().cbegin(), meta_.columns().cend(),
                           [&name](const basepb::ColumnInfo& col){ return col.name() == name; });
    if (it != meta_.columns().cend()) {
        return *it;
    } else {
        throw std::runtime_error(std::string("column name not found: ") + name);
    }
}

static dspb::ColumnInfo makeColumnInfo(const basepb::ColumnInfo& col) {
    dspb::ColumnInfo info;
    info.set_typ(col.sql_type().type());
    info.set_id(col.id());
    info.set_unsigned_(col.sql_type().unsigned_());
    return info;
}

dspb::ColumnInfo Table::GetColumnInfo(uint64_t id) const {
    return makeColumnInfo(GetColumn(id));
}

dspb::ColumnInfo Table::GetColumnInfo(const std::string& name) const {
    return makeColumnInfo(GetColumn(name));
}

void Table::AddColumn(const std::string& name, basepb::DataType type, bool is_pk) {
    auto col = meta_.add_columns();
    col->set_id(++last_col_id_);
    col->set_name(name);
    col->mutable_sql_type()->set_type(type);
    col->set_primary(is_pk ? 1 : 0);
}

std::unique_ptr<Table> CreateAccountTable(uint32_t tid) {
    std::unique_ptr<Table> t(new Table("account", tid));
    t->AddColumn("id", basepb::BigInt, true);
    t->AddColumn("name", basepb::Varchar);
    t->AddColumn("balance", basepb::BigInt);
    return t;
}

std::unique_ptr<Table> CreatePersonTable(uint32_t tid) {
    std::unique_ptr<Table> t(new Table("person", tid));
    t->AddColumn("id", basepb::BigInt, true);
    t->AddColumn("name", basepb::Varchar);
    t->AddColumn("age", basepb::SmallInt);
    t->AddColumn("height", basepb::Double);
    return t;
}


std::unique_ptr<Table> CreateUserTable(uint32_t tid) {
    std::unique_ptr<Table> t(new Table("user", tid));
    t->AddColumn("user_name", basepb::Varchar, true);
    t->AddColumn("pass_word", basepb::Varchar);
    t->AddColumn("real_name", basepb::Varchar);
    return t;
}

std::unique_ptr<Table> CreateAccountTable() {
    return CreateAccountTable(default_table_id);
}

std::unique_ptr<Table> CreatePersonTable() {
    return CreatePersonTable(default_person_table_id);
}

std::unique_ptr<Table> CreateUserTable() {
    return CreateUserTable(default_user_table_id);
}

std::unique_ptr<Table> CreateHashUserTable(uint32_t tid) {
    std::unique_ptr<Table> t(new Table("user", tid));
    t->AddColumn("h", basepb::Int, true);
    t->AddColumn("user_name", basepb::Varchar, true);
    t->AddColumn("pass_word", basepb::Varchar);
    t->AddColumn("real_name", basepb::Varchar);
    return t;
}

std::unique_ptr<Table> CreateHashUserTable() {
    return CreateHashUserTable(default_hashuser_table_id);
}

} /* namespace helper */
} /* namespace test */
} /* namespace jimkv */

File: jimsql/jim-sdk/jim-core/src/main/java/io/jimdb/core/model/meta/Table.java
/*
 * Copyright 2019 The JIMDB Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
package io.jimdb.core.model.meta;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.jimdb.common.exception.DBException;
import io.jimdb.common.exception.ErrorCode;
import io.jimdb.common.exception.ErrorModule;
import io.jimdb.pb.Basepb.StoreType;
import io.jimdb.pb.Metapb.ColumnInfo;
import io.jimdb.pb.Metapb.IndexInfo;
import io.jimdb.pb.Metapb.MetaState;
import io.jimdb.pb.Metapb.TableInfo;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;

/**
 * @version V1.0
 */
@SuppressFBWarnings({ "FCCD_FIND_CLASS_CIRCULAR_DEPENDENCY", "EI_EXPOSE_REP" })
public final class Table {
  private final Integer id;
  private final TableInfo tableInfo;
  private final Catalog catalog;

  private final Column[] primarys;
  private final Column[] readableColumns;
  private final Column[] writableColumns;

  private Index primaryIndex;
  private final Index[] readableIndices;
  private final Index[] writableIndices;
  private final Index[] deletableIndices;

  private final Map<String, Column> readableColumnMap;
  private final Map<String, Column> writableColumnMap;

  public Table(Catalog catalog, TableInfo tableInfo) {
    List<ColumnInfo> tableColumns = tableInfo.getColumnsList() == null ? Collections.EMPTY_LIST : tableInfo.getColumnsList();
    List<Integer> tablePrimarys = tableInfo.getPrimarysList() == null ? Collections.EMPTY_LIST : tableInfo.getPrimarysList();
    List<IndexInfo> tableIndex = tableInfo.getIndicesList() == null ? Collections.EMPTY_LIST : tableInfo.getIndicesList();

    this.id = tableInfo.getId();
    this.tableInfo = tableInfo;
    this.catalog = catalog;
    this.primarys = new Column[tablePrimarys.size()];
    this.readableColumnMap = tableColumns.size() == 0 ? Collections.EMPTY_MAP : new HashMap<>(tableColumns.size(), 1.0F);
    this.writableColumnMap = tableColumns.size() == 0 ? Collections.EMPTY_MAP : new HashMap<>(tableColumns.size(), 1.0F);

    int readOffset = 0;
    int writeOffset = 0;
    for (ColumnInfo columnInfo : tableColumns) {
      MetaState state = columnInfo.getState();
      if (state == MetaState.Absent || state == MetaState.DeleteOnly || state == MetaState.DeleteReorg) {
        continue;
      }

      if (state == MetaState.Public) {
        this.readableColumnMap.put(columnInfo.getName().toLowerCase(), new Column(this, columnInfo, readOffset++));
      }
      this.writableColumnMap.put(columnInfo.getName().toLowerCase(), new Column(this, columnInfo, writeOffset++));
    }
    List<Column> readableList = new ArrayList<>(this.readableColumnMap.values());
    List<Column> writableList = new ArrayList<>(this.writableColumnMap.values());
    Collections.sort(readableList, (c1, c2) -> {
      if (c1.getOffset() == c2.getOffset()) {
        return 0;
      }
      return c1.getOffset() > c2.getOffset() ? 1 : -1;
    });
    Collections.sort(writableList, (c1, c2) -> {
      if (c1.getOffset() == c2.getOffset()) {
        return 0;
      }
      return c1.getOffset() > c2.getOffset() ? 1 : -1;
    });
    this.readableColumns = readableList.toArray(new Column[0]);
    this.writableColumns = writableList.toArray(new Column[0]);

    for (int i = 0; i < tablePrimarys.size(); i++) {
      int colId = tablePrimarys.get(i);
      for (Column column : this.readableColumns) {
        if (column.getId() == colId) {
          this.primarys[i] = column;
          break;
        }
      }
    }

    Map<Integer, Index> readableIndexMap = tableIndex.isEmpty() ? Collections.EMPTY_MAP : new HashMap<>(tableIndex.size(), 1.0F);
    Map<Integer, Index> writableIndexMap = tableIndex.isEmpty() ? Collections.EMPTY_MAP : new HashMap<>(tableIndex.size(), 1.0F);
    Map<Integer, Index> deletableIndexMap = tableIndex.isEmpty() ? Collections.EMPTY_MAP : new HashMap<>(tableIndex.size(), 1.0F);
    for (IndexInfo indexInfo : tableIndex) {
      MetaState state = indexInfo.getState();
      if (state == MetaState.Absent) {
        continue;
      }

      Index index = new Index(this, indexInfo);
      if (index.isPrimary()) {
        primaryIndex = index;
      }
      if (state == MetaState.Public) {
        readableIndexMap.put(indexInfo.getId(), index);
      }
      if (state != MetaState.DeleteOnly && state != MetaState.DeleteReorg) {
        writableIndexMap.put(indexInfo.getId(), index);
      }
      deletableIndexMap.put(indexInfo.getId(), index);
    }
    List<Index> readIndexList = new ArrayList<>(readableIndexMap.values());
    List<Index> writeIndexList = new ArrayList<>(writableIndexMap.values());
    List<Index> deleteIndexList = new ArrayList<>(deletableIndexMap.values());
    this.sortIndex(readIndexList);
    this.sortIndex(writeIndexList);
    this.sortIndex(deleteIndexList);
    this.readableIndices = readIndexList.toArray(new Index[0]);
    this.writableIndices = writeIndexList.toArray(new Index[0]);
    this.deletableIndices = deleteIndexList.toArray(new Index[0]);
  }

  private void sortIndex(List<Index> indices) {
    Collections.sort(indices, (c1, c2) -> {
      if (c1.isPrimary()) {
        return -1;
      }
      if (c2.isPrimary()) {
        return 1;
      }

      if (c1.isUnique() && !c2.isUnique()) {
        return -1;
      }
      if (!c1.isUnique() && c2.isUnique()) {
        return 1;
      }

      return c1.getId().intValue() >= c2.getId().intValue() ? 1 : -1;
    });
  }

  public Integer getId() {
    return id;
  }

  public Integer getCatalogId() {
    return catalog.getId();
  }

  public Catalog getCatalog() {
    return catalog;
  }

  public String getName() {
    return tableInfo.getName();
  }

  public StoreType getEngine() {
    return tableInfo.getType();
  }

  public String getComment() {
    return tableInfo.getComment();
  }

  public Column[] getPrimary() {
    return primarys;
  }

  public Index getPrimaryIndex() {
    return primaryIndex;
  }

  public Column[] getReadableColumns() {
    return readableColumns;
  }

  public Column[] getWritableColumns() {
    return writableColumns;
  }

  public Index[] getReadableIndices() {
    return readableIndices;
  }

  public Index[] getWritableIndices() {
    return writableIndices;
  }

  public Index[] getDeletableIndices() {
    return deletableIndices;
  }

  public Column getReadableColumn(String name) {
    Column column = readableColumnMap.get(name.toLowerCase());
    if (column == null) {
      throw DBException.get(ErrorModule.PARSER, ErrorCode.ER_BAD_FIELD_ERROR, name, tableInfo.getName());
    }
    return column;
  }

  public Column getWritableColumn(String name) {
    Column column = writableColumnMap.get(name.toLowerCase());
    if (column == null) {
      throw DBException.get(ErrorModule.PARSER, ErrorCode.ER_BAD_FIELD_ERROR, name, tableInfo.getName());
    }
    return column;
  }

  public static Table getStatsTable() {
    return null;
//    Table statsTable = new Table();
//    statsTable.setId(904351);
//    statsTable.setName("statsTable");
//    statsTable.setDbId(904306);
//    statsTable.setDbName("stats");
//    JimColumn[] columns = new JimColumn[14];
//
//    JimColumn idColumn = new JimColumn(1L, "id", DataType.BigInt, 0);
//    //idColumn.setPrimaryKey(1);
//    columns[0] = idColumn;
//
//    columns[1] = new JimColumn(2L, "isIndex", DataType.Binary, 1);
//    columns[2] = new JimColumn(3L, "rowCount", DataType.BigInt, 2);
//    columns[3] = new JimColumn(4L, "hId", DataType.BigInt, 3);
//    columns[4] = new JimColumn(5L, "hNdv", DataType.BigInt, 4);
//    columns[5] = new JimColumn(6L, "hNullCount", DataType.BigInt, 5);
//    columns[6] = new JimColumn(7L, "hTotalCount", DataType.Double, 6);
//    columns[7] = new JimColumn(8L, "hBuckets", DataType.Varchar, 7);
//    columns[8] = new JimColumn(9L, "cRowCount", DataType.Varchar, 8);
//    columns[9] = new JimColumn(10L, "cColumnCount", DataType.Varchar, 9);
//    columns[10] = new JimColumn(11L, "cHashTables", DataType.Varchar, 10);
//    columns[11] = new JimColumn(12L, "cDefaultValue", DataType.BigInt, 11);
//    columns[12] = new JimColumn(13L, "cNdv", DataType.BigInt, 12);
//    columns[13] = new JimColumn(14L, "cTotalCount", DataType.BigInt, 13);
//
//    statsTable.setColumns(columns);
//    return statsTable;
  }
}
